Add actual/expected hit count stats to stats dialog

jostephd opened the pull request:
Fixes #4066, see https://github.com/wesnoth/wesnoth/issues/4066#issuecomment-490973775 for an explanation.

Currently looks like this:

![2019-05-10-112557_1920x1080_scrot](https://user-images.githubusercontent.com/24784687/57524313-f8abe380-7316-11e9-96b4-96c696a446d5.png)

The screenshot shows the status after a single attack by a Silver Mage (9x4 magical) that hit 3/4.

jostephd commented on src/statistics.cpp:
```diff
@@ -376,6 +436,18 @@ void stats::read(const config& cfg)
+	}
+	if (const config &c = cfg.child("turn_by_cth_taken")) {
+		turn_by_cth_taken = read_by_cth_map(c);
+	}
```
I think we can remove `by_cth_inflicted`/`by_cth_taken` from serialization, and instead derive them from the `attacks` and `defends` members, which are read a few lines above:

https://github.com/wesnoth/wesnoth/blob/635401a307c29ce79b29034665e8be9b399755d3/src/statistics.hpp#L41-L50

With this change we'll be able to show the hit statistics for old savefiles too.

The `turn_by_cth_*` members should still be serialized.

<details>

<summary>
Savefile excerpt
</summary>

```
			[attacks]
				[sequence]
					1="s1001"
					2="s00"
					3="s11,s1,s10"
					7="s01"
					_num=40
				[/sequence]
				[sequence]
					2="s0101,s10"
					4="s11"
					5="s00,s01"
					_num=50
				[/sequence]
				[sequence]
					1="s001,s0010,s0100,s0110,s0111,s1011,s1100,s1101,s111"
					2="s10"
					4="s00"
					7="s11,s01"
					_num=60
				[/sequence]
				[sequence]
					1="s00,s11"
					_num=80
				[/sequence]
			[/attacks]
			[defends]
				[sequence]
					12="s"
					_num=0
				[/sequence]
				[sequence]
					1="s00"
					_num=40
				[/sequence]
				[sequence]
					1="s10"
					_num=50
				[/sequence]
				[sequence]
					1="s1,s10,s101"
					3="s00"
					5="s"
					9="s11"
					11="s01"
					_num=60
				[/sequence]
				[sequence]
					1="s01,s11"
					_num=80
				[/sequence]
			[/defends]
```

The first block in the `defends` section may need special casing, but other than that it looks straightforward enough.

</details>

jostephd commented on src/gui/dialogs/statistics_dialog.cpp:
```diff
@@ -135,32 +156,149 @@ void statistics_dialog::add_damage_row(
+	}
+
+	std::ostringstream str, str2;
+	str << overall_hits << " / " << expected_hits;
```
@Pentarctagon https://github.com/wesnoth/wesnoth/issues/4066#issuecomment-491360346

> Otherwise, I'd say to use some other separator than `/`. I can imagine there being a fair number of people confused about how 2 divided by 2.7 somehow ends up at 39%. Maybe display it as `2 (2.7)` instead? 17/25.2 is at least fairly straight forward as to how they are related to the -32%.

jostephd replied:
Quoting from [the forums](https://forums.wesnoth.org/viewtopic.php?f=12&t=49785&p=642417#p642417):

> > (Mawmoocn) For visual UI, instead of 80 / 83.3, it could be, 80 = 83.3, [80 = 83.3] [28.0%] [72.0%], [80 - 83.3 (28.0%)] [72.0%], [80 | 83.3 (28.0%)]. 

> Thanks. I wouldn't use = here because 80 is not equal to 83.3. I suppose | is an option but I don't think it's very intuitive. Hmm, maybe "80 (-3.3)" or "83.3 ⊨ 80"? :thinking:

jostephd replied:
@Pentarctagon Just to tie up loose ends, I assume this is resolved now that the dialog shows two percentages instead of one?

jyrkive commented on src/gui/dialogs/statistics_dialog.cpp:
```diff
@@ -154,34 +156,35 @@ void statistics_dialog::add_damage_row(
 	str.str("");
+	str << (((dsa < 0) ^ (expected < 0)) ? "" : "+")
+		<< (expected == 0 ? 0 : 100 * dsa / expected) << '%';
+	item["label"] = str.str() + spacer;
```
Just make the spacer a part of the string stream so that you don't need to append it afterwards.

jyrkive commented on src/gui/dialogs/statistics_dialog.cpp:
```diff
@@ -154,34 +156,35 @@ void statistics_dialog::add_damage_row(
+		str.str("");
+		str << (((dst < 0) ^ (turn_expected < 0)) ? "" : "+")
+			<< (turn_expected == 0 ? 0 : 100 * dst / turn_expected) << '%';
+		item["label"] = str.str() + spacer;
```
Same here.

jyrkive commented on src/gui/dialogs/statistics_dialog.cpp:
```diff
@@ -154,34 +156,35 @@ void statistics_dialog::add_damage_row(
-// Return the string to use in the "Hits" table, showing actual and expected number of hits.
-static std::string tally(const std::map<int, struct statistics::stats::by_cth_t>& by_cth)
+// Return the strings to use in the "Hits" table, showing actual and expected number of hits.
+static std::pair<std::string, std::string> tally(const std::map<int, struct statistics::stats::by_cth_t>& by_cth)
```
Please create a named struct instead of a pair.

jostephd replied:
Created in c6e7bb5e0bb.

jyrkive commented on src/gui/dialogs/statistics_dialog.cpp:
```diff
@@ -275,14 +277,23 @@ void statistics_dialog::add_hits_row(
+	pair = tally(by_cth);
+	item["label"] = pair.first;
 	data.emplace("hits_overall", item);
+	item["label"] = pair.second + spacer;
```
Likewise, make the spacer a part of the string so that you don't need to append it afterwards.

jyrkive commented on src/statistics.cpp:
```diff
@@ -207,6 +207,34 @@ static config write_by_cth_map(const std::map<int, struct statistics::stats::by_
 }
 
+static void merge_battle_result_maps(stats::battle_result_map& a, const stats::battle_result_map& b);
+static std::map<int, struct statistics::stats::by_cth_t> read_by_cth_map_from_battle_result_maps(const statistics::stats::battle_result_map& attacks, const statistics::stats::battle_result_map& defends)
```
Remove `struct`.

jyrkive commented on src/statistics.cpp:
```diff
@@ -207,6 +207,34 @@ static config write_by_cth_map(const std::map<int, struct statistics::stats::by_
+static void merge_battle_result_maps(stats::battle_result_map& a, const stats::battle_result_map& b);
+static std::map<int, struct statistics::stats::by_cth_t> read_by_cth_map_from_battle_result_maps(const statistics::stats::battle_result_map& attacks, const statistics::stats::battle_result_map& defends)
+{
+	std::map<int, struct statistics::stats::by_cth_t> m;
```
Remove `struct`.

jyrkive commented on src/statistics.cpp:
```diff
@@ -207,6 +207,34 @@ static config write_by_cth_map(const std::map<int, struct statistics::stats::by_
+			const int occurrences = j.second;
+			unsigned int misses = std::count(res.begin(), res.end(), '0');
+			unsigned int hits = std::count(res.begin(), res.end(), '1');
+			if(misses + hits == 0)
```
Please use braces.

jostephd replied:
Done in 14a7139c018af0a5ccea6a110ab213a4ff6b17a7 along with the other smaller changes.

jostephd commented on src/gui/dialogs/statistics_dialog.cpp:
```diff
@@ -164,19 +183,83 @@ void statistics_dialog::add_damage_row(
 	std::ostringstream str;
 	str << overall_hits << " / " << expected_hits;
+
+	// Compute the a priori probability of this actual result, by simulating many attacks against a single defender.
```
> It's even more confusing that depending on the number of hits, the percentage may the probability of _this exact result_, the probability of _this or more hits_, or the probability of _this or less hits_, i.e. completely different things.

"This exact result" is a rare case since the expected value is usually not an integer. Perhaps we could display an em dash in that case? There is precedent for that: the "Gold cost" column in the "Advancements" row shows an em dash.

The other two cases are exactly the same: they show the _a priori_ probability of the actual result. Isn't that what players would want to know? When a player hits above average, they would ask "How lucky was I"; when a player hits below average, they would ask "How unlucky was I".

There is a tooltip that explains this briefly: https://github.com/wesnoth/wesnoth/pull/4070/commits/67847506d4c1d1dae1eb19f820fea02c2fb6e0c6#diff-6f98c6223f5ee298110c1deef256831eR562

I suspected a tooltip wouldn't be enough but wasn't sure what else to do. I suppose I could add a static label to the dialog, or a new page to the in-game help section, to explain what the table shows? Would that address the issue?

What do you think of showing "probability of having this many hits or more"? This would make it so if a player was very _lucky_ the percentage shown would be small, but if the player was very _unlucky_ the percentage shown would be large. I honestly don't know if I like this better than the currently-implemented behavior. Treating the two extreme cases differently like this feels wrong somehow. If a unit has 50% chance to hit, then "very lucky" and "very unlucky" are _both_ rare/extreme cases, so showing a small percentage in both of these cases feels appropriate.

jyrkive replied:
I think always showing "probability of this many hits or more" would be significantly more consistent, and thus a huge improvement.

jostephd replied:
:thinking: 

I see your point that showing different things in the same label is poor UI design, but I stand by my point that "very lucky" and "very unlucky" should be presented similarly. For example, there is no particular reason to prefer implementing "always show 'probability of this many hits or more'" over implementing "always show 'probability of this many hits or _less_'".

That being said, I'm not sure what change to make. I just [asked the forums](https://forums.wesnoth.org/viewtopic.php?f=12&t=49785&p=642368#p642368) for ideas. I'd like to address @Pentarctagon's feedback at the same time.

jostephd replied:
I've just pushed two commits (e70c1ed863ee4565689ee174ff4224354bf0ee65 and 1dfe728e04caea865ae17ceb998051da8da58957). The log messages explain what they do. Here's what they look like:

![2019-05-19-223318_1920x1080_scrot](https://user-images.githubusercontent.com/24784687/57989263-cf622480-7a87-11e9-815c-48d535f243b7.png)
![2019-05-19-223411_1920x1080_scrot](https://user-images.githubusercontent.com/24784687/57989264-cf622480-7a87-11e9-9e38-d434d1f67f0b.png)

The screenshots show (especially in the "Overall" column) how when the RNG was on your side the numbers are red and when the RNG was on the enemy's side the numbers are green.

jyrkive replied:
I think it would work better with some simplification. You could remove the "probability of actual result" and the coloring, they're both redundant.

jostephd replied:
Thanks for the suggestions. I'll remove "probability of actual result". (I might patch it back in in my local branch, though.) As to coloring, I think it helps, because whether large numbers are good or bad varies. In the screenshot, 63.9% and 71.5% are colored in "green to red", and 24.6% and 18.4% are colored in "red to green". For example, if the stats were 5%/0%/95% in both taken and inflicted, the first row would be green and the second one red. I think it would be poor UI to just show all four of these figures in white.

On the other hand, I decided against coloring the percentages in the damage table, because in that case I do think coloring is redundant. Because of regression to the mean, the percentage values in that table are always around 0%, and when they aren't, the sample size is too small anyway.

jostephd commented on src/statistics.cpp:
```diff
@@ -208,6 +230,14 @@ static void merge_battle_result_maps(stats::battle_result_map& a, const stats::b
 	}
 }
 
+static void merge_cth_map(std::map<int, struct statistics::stats::by_cth_t>& a, const std::map<int, struct statistics::stats::by_cth_t>& b)
```
I've defined a type alias using `typedef` to be consistent with other member types of `statistics::stats` and used it throughout, which addressed most of your comments about removing the `struct` keyword.

jyrkive replied:
For the record, `typedef` is more common in existing code because the using aliases were only introduced in C++11. However, I'm also fine with `typedef`.

jostephd commented on src/gui/dialogs/statistics_dialog.cpp:
```diff
@@ -164,19 +183,83 @@ void statistics_dialog::add_damage_row(
+			config attacker_cfg(
+				"id", "statistics_dialog_dummy_attacker" + std::to_string(cth),
+				"hide_help", "yes",
+				"do_not_list", "yes",
```
Done.  I also found another instance but did not change it since I was not 100% sure the change would be correct there: https://github.com/wesnoth/wesnoth/blob/40fe145213b42a56df68e2c63f16714751bdc39f/src/statistics.cpp#L621

jyrkive replied:
The change would indeed be correct there as well.

jostephd replied:
Done: 39bdda9d256

jostephd commented on src/gui/dialogs/statistics_dialog.cpp:
```diff
@@ -164,19 +183,83 @@ void statistics_dialog::add_damage_row(
+		if(overall_hits == expected_hits) {
+			probability = chance_of_exactly_N_hits(overall_hits);
+		}
+		else if (overall_hits > expected_hits) {
```
Changed. We have plenty of code following both styles, but I'll put `else` on the same line in new code.

jostephd commented on src/gui/dialogs/statistics_dialog.cpp:
```diff
@@ -161,6 +161,45 @@ void statistics_dialog::add_damage_row(
+		expected_hits += (cth * 0.01) * i.second.strikes;
+	}
+
+	str.str("");
```
That line had already been removed in a later commit: a764a336352a12fd546d1c9f2edc9e3c8356d87b

jostephd commented on src/statistics.cpp:
```diff
@@ -194,6 +194,28 @@ static stats::battle_result_map read_battle_result_map(const config& cfg)
+			"key", i.first,
+			"value", i.second.write()
+		);
+		res.add_child("keyvalue", child);
```
Before I make the change, is there generic `std::map`-to-WML serialization code I should be using?

jyrkive replied:
I'm not aware of any. Such code would be tricky to write because `std::map` is a template.

jostephd replied:
Alright. I changed the strings as you suggested: c684d72f854

jostephd commented:
Thanks for the thorough review! I've addressed all your comments except https://github.com/wesnoth/wesnoth/pull/4070#discussion_r283120928, which I will address shortly. I had one question, see https://github.com/wesnoth/wesnoth/pull/4070#discussion_r283119711.

jyrkive commented on src/statistics.cpp:
```diff
@@ -758,7 +758,7 @@ config write_stats()
 void write_stats(config_writer &out)
 {
-	out.write_key_val("mid_scenario", mid_scenario ? "yes" : "no");
+	out.write_key_val("mid_scenario", mid_scenario ? true : false);
```
Because `mid_scenario` is already a boolean, the ternary expression is unnecessary.

gfgtdf commented:
The empty brackets I'm the title looks strange, i don't know which text idiosyncrasies good there but maybe we should not display then when that text is empty (I know that this is not caused by your PR, just noticed it in the image)

jostephd commented:
> The empty brackets I'm the title looks strange, i don't know which text idiosyncrasies good there but maybe we should not display then when that text is empty (I know that this is not caused by your PR, just noticed it in the image)

#4080

jostephd commented on data/gui/window/statistics_dialog.cfg:
```diff
@@ -155,6 +167,18 @@
+					[label]
+						definition = "default_small"
+						label = "" # percentage
+						linked_group = "detail"
```
The linked_group line might be wrong.

jostephd commented on data/gui/window/statistics_dialog.cfg:
```diff
@@ -143,6 +143,18 @@
+					[label]
+						definition = "default_small"
+						label = "" # percentage
+						linked_group = "detail"
```
The linked_group line might be wrong.

jostephd commented on data/gui/window/statistics_dialog.cfg:
```diff
@@ -177,41 +177,105 @@
+                                                    definition = "default_small"
+                                                    tooltip = {ratio_tooltip}
+                                                [/label]
+                                            [/column]
```
Mawmoocn on the forums asks about vertical alignment between the hits and damage tables and between the percentage values in each table.

jostephd replied:
That has been fixed already, hasn't it?

jostephd commented:
Thank you for reviewing!

jostephd commented:
Rebased to squash/reorder a few changes and to trigger travis and appveyor.

Vultraz commented on data/gui/window/statistics_dialog.cfg:
```diff
@@ -108,9 +108,9 @@
 #enddef
 
-#define _GUI_DAMAGE_STATS_LIST
+#define _GUI_DAMAGE_STATS_LIST top_left_label overall_id colid_type colid_overall colid_this_turn ratio_tooltip percent_tooltip
```
Macro names should be uppercase.

jostephd replied:
Done in 46b468eda0f

Vultraz commented on data/gui/window/statistics_dialog.cfg:
```diff
@@ -176,42 +176,106 @@
+                                [column]
+                                    grow_factor = 1
+                                    border = "left"
+                                    border_size = -5
```
Having borders around grids isn't usually done except in very specific circumstances, and negative border sizes are not supported anyway. Please remove.

jostephd replied:
Per discussion on discord we'll address this after merge.

@Vultraz Thanks for the review! I think I've addressed everything. I've also rebased to master.

jostephd replied:
I changed the -5 to 0 and it still looks correct. I'll merge.

Vultraz commented on src/gui/dialogs/statistics_dialog.cpp:
```diff
@@ -133,34 +151,169 @@ void statistics_dialog::add_damage_row(
-		<< (expected == 0 ? 0 : 100 * dsa / expected) << '%';
-
-	item["label"] = str.str();
+	const auto& damage_str = [shift](long long damage, long long expected) {
```
Shouldn't be a reference. I'm not even sure what this would technically reference...

jostephd replied:
`damage_str` would reference an anonymous lambda function that exists until the end of the block. https://en.cppreference.com/w/cpp/language/reference_initialization#Lifetime_of_a_temporary

I've removed the `&`, but I think the code is basically equivalent with and without it.

Vultraz commented on src/gui/dialogs/statistics_dialog.cpp:
```diff
@@ -133,34 +151,169 @@ void statistics_dialog::add_damage_row(
 	data.emplace("damage_overall", item);
 
-	str.str("");
+	const auto& percent_str = [](long long dsx, long long expected) {
```
Shouldn't be a reference.

jostephd replied:
21dc2351a27

Vultraz commented on src/gui/dialogs/statistics_dialog.cpp:
```diff
@@ -133,34 +151,169 @@ void statistics_dialog::add_damage_row(
 }
 
+// Custom type to allow tally() to return two values.
+struct hitrate_table_element {
```
Formatting, leading struct brace goes on its own line.

jostephd replied:
e9d7b248c6c6b94ff340d831549638cef51b08f4

Vultraz commented on src/gui/dialogs/statistics_dialog.cpp:
```diff
@@ -133,34 +151,169 @@ void statistics_dialog::add_damage_row(
+			unit_type attacker_type(attacker_cfg);
+			unit_types.build_unit_type(attacker_type, unit_type::BUILD_STATUS::FULL);
+
+			attack_ptr attack = std::make_shared<attack_type>(config(
```
Just use `auto` here instead of `attack_ptr` since you're using `make_shared`.

jostephd replied:
e9d7b248c6c6b94ff340d831549638cef51b08f4

Vultraz commented on src/gui/dialogs/statistics_dialog.cpp:
```diff
@@ -133,34 +151,169 @@ void statistics_dialog::add_damage_row(
+		}
+
+		const std::vector<double>& final_hp_dist = current_defender->hp_dist;
+		const auto& chance_of_exactly_N_hits = [&final_hp_dist](int n) { return final_hp_dist[final_hp_dist.size() - 1 - n]; };
```
Should not be a reference.

jostephd replied:
b89ba82335e

Vultraz commented on src/gui/dialogs/statistics_dialog.cpp:
```diff
@@ -133,34 +151,169 @@ void statistics_dialog::add_damage_row(
+			// Start of turn
+			str2 << "0%";
+		} else {
+			const auto& add_probability = [&str2](double probability, bool more_is_better) {
```
Should not be a reference.

jostephd replied:
21dc2351a2751ff0ac3c91f558e0450a429c0bfa

Vultraz commented on src/statistics.cpp:
```diff
@@ -194,6 +194,57 @@ static stats::battle_result_map read_battle_result_map(const config& cfg)
+{
+	config res;
+	for(const auto& i : m) {
+		const config child(
```
You don't need to create a temp config, object here. You should be able to just do
```cpp
res.add_child("hitrate_map_entry", { 
    "cth", i.first,
    "stats", i.second.write() }
);
```
if I remember correctly.

jostephd replied:
6e10caddfde

jostephd commented:
Rebased to master.

jostephd commented:
It looks like this:

![2019-06-11-085338_1920x1080_scrot](https://user-images.githubusercontent.com/24784687/59257988-a1a37200-8c26-11e9-886e-0c17cd78b9aa.png)

That's after one Silver Mage / Orcish Archer battle.
